File: algorithms/[1833]雪糕的最大数量/solution.ts
// topics = ["贪心算法"]

/**
 * 排序（贪心）
 * time O(n + C), space O(C), n 为 const是长度, C = 10^5
 */
function maxIceCream(costs: number[], coins: number): number {
  // 最贵雪糕的价格
  const maxV = 10 ** 5;
  // freq[i] 价格为 i 的雪糕的个数
  const freq = new Array(maxV + 1).fill(0);

  for (const c of costs) {
    freq[c] += 1;
  }

  let count = 0;
  for (let i = 1; i < maxV + 1; i += 1) {
    if (coins < i) break;
    const curCount = Math.min(freq[i], Math.floor(coins / i));
    coins -= i * curCount;
    count += curCount;
  }

  return count;
}

File: algorithms/[1833]雪糕的最大数量/solution.py
# topics = ["数组", "贪心算法"]

from typing import List


class Solution:
    def maxIceCream(self, costs: List[int], coins: int) -> int:
        """
        排序（贪心）
        time O(nlong), space O(logn), n 为 costs 长度
        """
        costs.sort()

        preSum = 0
        res = 0
        for c in costs:
            preSum += c
            if preSum > coins:
                return res
            res += 1

        return len(costs)
